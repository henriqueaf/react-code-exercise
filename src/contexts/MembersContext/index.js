import React, { createContext, useEffect, useState, useContext } from 'react';

import { useMembersReducer } from '../../reducers/MembersReducer';
import { setLoading, setMembers, setFilteredMembers, setCurrentPage, setSelectedSession } from '../../reducers/MembersReducer/actions';
import { getMembersCache } from '../../services/membersCache';

const HOUSE_MINIMUM_SESSION = 102;
const SENATE_MINIMUM_SESSION = 80;

const MembersContext = createContext({});

const MembersProvider = ({ children }) => {
  const [state, dispatch] = useMembersReducer();
  const {
    filteredMembers,
    membersPerPage,
    selectedChamber,
    selectedSession,
  } = state;

  const [minimumSession, setMinimumSession] = useState(SENATE_MINIMUM_SESSION);

  const calculateMinimumSession = () => {
    switch (selectedChamber) {
      case 'senate':
        return SENATE_MINIMUM_SESSION;
      case 'house':
        return HOUSE_MINIMUM_SESSION;
    }
  };

  const providerValue = {
    minimumSession,
    state,
    dispatch,
  };

  useEffect(() => {
    dispatch(setLoading(true));

    const url = `https://api.propublica.org/congress/v1/${selectedSession}/${selectedChamber}/members.json`;

    async function membersCache() {
      try {
        const responseMembers = await getMembersCache(url);
        dispatch(setMembers(responseMembers));
        dispatch(setFilteredMembers(responseMembers));
      } catch (error) {
        console.error(error.message);
        dispatch(setLoading(false));
      } finally {
        dispatch(setLoading(false));
        setMinimumSession(calculateMinimumSession());
      }
    }

    membersCache()
  }, [selectedChamber, selectedSession]);

  useEffect(() => {
    if (selectedChamber === 'house' && selectedSession < HOUSE_MINIMUM_SESSION)
      dispatch(setSelectedSession(HOUSE_MINIMUM_SESSION));
  }, [selectedChamber]);

  useEffect(() => {
    dispatch(setCurrentPage(0));
  }, [membersPerPage, filteredMembers]);

  return (
    <MembersContext.Provider value={providerValue}>
      {children}
    </MembersContext.Provider>
  );
};

const useMembersContext = () => useContext(MembersContext);

export { useMembersContext, MembersProvider, MembersContext };
