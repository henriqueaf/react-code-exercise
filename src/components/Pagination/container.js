import React from 'react';
import { useMembersContext } from '../../contexts/MembersContext';

import Pagination from './index';

const Container = () => {
  const {
    state: {
      filteredMembers,
      membersPerPage,
      currentPage
    },
    dispatch
  } = useMembersContext();

  return (
    <Pagination
      filteredMembers={filteredMembers}
      membersPerPage={membersPerPage}
      currentPage={currentPage}
      dispatch={dispatch}
    />
  )
}

export default Container;
