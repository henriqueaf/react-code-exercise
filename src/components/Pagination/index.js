import React from 'react';
import { Pagination, Form, Row, Col } from 'react-bootstrap';
import isNumber from 'lodash/isNumber';

import { useMembersContext } from '../../contexts/MembersContext';

export default () => {
  const {
    filteredMembers,
    membersPerPage,
    setMembersPerPage,
    currentPage,
    setCurrentPage
  } = useMembersContext();

  const pages = Math.ceil(filteredMembers.length / membersPerPage);
  const pagesArray = [];

  for (let number = 0; number < pages; number++) {
    pagesArray.push(
      <Pagination.Item key={number} active={number === currentPage} onClick={() => setCurrentPage(number)}>
        {number + 1}
      </Pagination.Item>
    );
  }

  return (
    <Row>
      <Col>
        <Pagination>
          <Pagination.First onClick={() => setCurrentPage(0)} />
          <Pagination.Prev onClick={() => currentPage > 0 && setCurrentPage(currentPage - 1)} />
          {pagesArray}
          <Pagination.Next onClick={() => currentPage < (pages - 1) && setCurrentPage(currentPage + 1)} />
          <Pagination.Last onClick={() => setCurrentPage(pages - 1)} />
        </Pagination>
      </Col>

      <Col sm={1}>
        <Form.Select value={membersPerPage} onChange={({ target: { value } }) => setMembersPerPage(value)}>
          <option>Select members per page</option>
          <option value={10}>10</option>
          <option value={20}>20</option>
          <option value={30}>30</option>
        </Form.Select>
      </Col>
    </Row>
  )
};
