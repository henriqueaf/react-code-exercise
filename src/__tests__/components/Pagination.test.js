import {render, fireEvent, waitFor} from '@testing-library/react';
import { MembersProvider } from '../../contexts/MembersContext';
import Pagination from '../../components/Pagination/container';

const member = {
  id: 'A00123',
  suffix: 'Sr.',
  first_name: 'Lamar',
  middle_name: 'Adam',
  last_name: 'Frank',
  twitter_account: 'lamartwt',
  facebook_account: 'lamarfc',
  youtube_account: 'lamaryoutube',
  party: 117,
  district: 'LA',
  state: 'LA',
  next_election: 2022
}

const mount = () => {
  return render(
    <MembersProvider>
      <Pagination />
    </MembersProvider>
  );
};

beforeEach(() => {
  // eslint-disable-next-line no-undef
  global.fetch = jest.fn(() =>
    Promise.resolve({
      json: () => Promise.resolve({
        results: [{
          members: Array(23).fill(member)
        }]
      })
    })
  );
});

afterEach(() => {
  jest.clearAllMocks();
});

describe('Pagination', () => {
  it('Matches snapshot', async () => {
    let container;
    await waitFor(() => {
      container = mount().container;
    });
    expect(container).toMatchSnapshot();
  });

  it('Set members per page', async () => {
    const {findByRole, findAllByRole} = await waitFor(() => mount());

    let pagesLinks = await findAllByRole('listitem');

    expect(pagesLinks.length).toBe(7);

    const membersPerPageCombobox = await findByRole('combobox');
    fireEvent.change(membersPerPageCombobox, { target: { value: 30 } });

    pagesLinks = await findAllByRole('listitem');
    expect(pagesLinks.length).toBe(5);
  });

  it('First page active', async () => {
    const {findByText} = mount();
    const currentPageButton = await findByText('1');

    expect(currentPageButton.parentElement.className).toEqual('page-item active');
  });

  it('Set current page', async () => {
    const {findByText} = mount({filteredMembers: Array(23).fill(member)});
    const lastPageButton = await findByText('Last');

    fireEvent.click(lastPageButton);

    const currentPageButton = await findByText('3');
    expect(currentPageButton.parentElement.className).toEqual('page-item active');
  });
})
