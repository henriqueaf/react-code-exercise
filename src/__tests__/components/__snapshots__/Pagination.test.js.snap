// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Matches snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col"
    >
      <ul
        class="pagination"
      >
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            <span
              aria-hidden="true"
            >
              «
            </span>
            <span
              class="visually-hidden"
            >
              First
            </span>
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            <span
              aria-hidden="true"
            >
              ‹
            </span>
            <span
              class="visually-hidden"
            >
              Previous
            </span>
          </a>
        </li>
        <li
          class="page-item active"
        >
          <span
            class="page-link"
          >
            1
            <span
              class="visually-hidden"
            >
              (current)
            </span>
          </span>
        </li>
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            2
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            3
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            <span
              aria-hidden="true"
            >
              ›
            </span>
            <span
              class="visually-hidden"
            >
              Next
            </span>
          </a>
        </li>
        <li
          class="page-item"
        >
          <a
            class="page-link"
            href="#"
            role="button"
          >
            <span
              aria-hidden="true"
            >
              »
            </span>
            <span
              class="visually-hidden"
            >
              Last
            </span>
          </a>
        </li>
      </ul>
    </div>
    <div
      class="col-sm-1"
    >
      <select
        class="form-select"
      >
        <option>
          Select members per page
        </option>
        <option
          value="10"
        >
          10
        </option>
        <option
          value="20"
        >
          20
        </option>
        <option
          value="30"
        >
          30
        </option>
      </select>
    </div>
  </div>
</div>
`;
